core/src/main/scala/spinoco/fs2/cassandra/internal/ctype/CollectionCType.scala

```diff
@@ -0,0 +1,149 @@
+
+    val elementCodec = codecs.elementCodec(CType[A].cqlCodec(protocolVersion))
+
+
+    new Codec[C[A]] {
+      def encode(value: C[A]): Attempt[BitVector] = {
+        @tailrec
+        def go(remains: C[A], acc: BitVector): Attempt[BitVector] = {
+          CollectionType[C].uncons1(remains) match {
+            case None => Attempt.successful(acc)
+            case Some((head, tail)) =>
+              elementCodec.encode(head) match {
+                case Attempt.Successful(encodedBits) => go(tail, acc ++ encodedBits)
+                case Attempt.Failure(err) => Attempt.failure(Err.General(s"Failed to encode element: $head", err.message +: err.context))
+              }
+          }
+        }
+
+        val elementCount = BitVector.fromInt(CollectionType[C].sizeOf(value))
+        go(value, elementCount)
+      }
+
+      def sizeBound: SizeBound = SizeBound.unknown
+
+      def decode(bits: BitVector): Attempt[DecodeResult[C[A]]] = {
+        if (bits.isEmpty) Attempt.Successful(DecodeResult(CollectionType[C].zero, bits))
+        else {
+          scodec.codecs.int32.decode(bits).flatMap { case DecodeResult(count, bits) =>
+            def go(remains: Int, remainsBits: BitVector, acc: C[A]): Attempt[DecodeResult[C[A]]] = {
+              if (remains > 0) {
+                elementCodec.decode(remainsBits) match {
+                  case Attempt.Successful(DecodeResult(element, rest)) =>
+                    go(remains - 1, rest, CollectionType[C].append(acc, element))
+                  case Attempt.Failure(err) =>
+                    Attempt.failure(Err.General(s"Failed to decode element at ${count - remains}", err.message +: err.context))
+                }
+              } else {
+                Attempt.successful(DecodeResult(acc, remainsBits))
+              }
+            }
+            go(count, bits, CollectionType[C].zero)
+          }
+        }
+      }
+    }
+  }
+
+  /**
+    * Formats collection with CQL syntax [elem1, elem2, elem3]
+    */
+  def format[C[_] : CollectionType, A : CType](ca: C[A]): Attempt[String] = {
+    def go(remains: C[A], acc: String): Attempt[String] = {
+      CollectionType[C].uncons1(remains) match {
+        case None => Attempt.Successful(s"[$acc]")
+        case Some((head, tail)) =>
+          CType[A].format(head) match {
+            case Attempt.Successful(formatted) =>
+              if (acc.isEmpty) go(tail, formatted)
+              else go(tail, acc +  "," + formatted)
+
+            case Attempt.Failure(err) =>
+              Attempt.failure(Err.General(s"Failed to format element: $head", err.message +: err.context))
+
+          }
+      }
+    }
+    go(ca, "")
+  }
+
+  /**
+    * Simple parser of the CQL collection value from the string
+    * @param cql             Source CQL string
+    * @param collectionType  Type of the collection
+    * @tparam C
+    * @tparam A
+    * @return
+    */
+  def parse[C[_] : CollectionType, A : CType](cql: String): Attempt[C[A]] = {
```

Comment:
cannot we borow that from native codec?

Reply:
O, I see it is not new code - just code moved from types.scala.